firecrown/likelihood/gauss_family/gauss_family.py

```diff
@@ -98,28 +115,91 @@ def read(self, sacc_data: sacc.Sacc) -> None:
             raise RuntimeError(msg)
 
         covariance = sacc_data.covariance.dense
+
+        indices_list = []
+        data_vector_list = []
         for stat in self.statistics:
             stat.read(sacc_data)
+            if stat.statistic.sacc_indices is None:
+                raise RuntimeError(
+                    f"The statistic {stat.statistic} has no sacc_indices."
+                )
+            indices_list.append(stat.statistic.sacc_indices.copy())
+            data_vector_list.append(stat.statistic.get_data_vector())
 
-        indices_list = [s.statistic.sacc_indices.copy() for s in self.statistics]
         indices = np.concatenate(indices_list)
+        data_vector = np.concatenate(data_vector_list)
         cov = np.zeros((len(indices), len(indices)))
 
         for new_i, old_i in enumerate(indices):
             for new_j, old_j in enumerate(indices):
                 cov[new_i, new_j] = covariance[old_i, old_j]
 
+        self.data_vector = data_vector
+        self.cov_index_map = {old_i: new_i for new_i, old_i in enumerate(indices)}
         self.cov = cov
         self.cholesky = scipy.linalg.cholesky(self.cov, lower=True)
         self.inv_cov = np.linalg.inv(cov)
 
         self.state = State.READY
 
+    def make_realization(self, sacc_data: sacc.Sacc, strict=True) -> sacc.Sacc:
+        assert (
+            self.state == State.UPDATED
+        ), "update() must be called before get_theory_vector()"
+
+        if not self.computed_theory_vector:
+            raise RuntimeError(
+                "The theory vector has not been computed yet. "
+                "Call compute_theory_vector first."
+            )
+
+        new_sacc = sacc_data.copy()
+
+        sacc_indices_list = []
+        predictions_list = []
+        for stat in self.statistics:
+            assert stat.statistic.sacc_indices is not None
+            sacc_indices_list.append(stat.statistic.sacc_indices.copy())
+            predictions_list.append(stat.statistic.get_theory_vector())
+
+        sacc_indices = np.concatenate(sacc_indices_list)
+        predictions = np.concatenate(predictions_list)
+        assert len(sacc_indices) == len(predictions)
+
+        if strict:
+            if set(sacc_indices.tolist()) != set(sacc_data.indices()):
+                raise RuntimeError(
+                    "The predicted data does not cover all the data in the "
+                    "sacc object. To write only the calculated predictions, "
+                    "set strict=False."
+                )
+
+        # Adding Gaussian noise defined by the covariance matrix.
+        assert self.cholesky is not None
+        predictions += np.dot(self.cholesky, np.random.randn(len(predictions)))
```

Comment:
Adding noise should be optional. A common use case is that we want a noise-free theory prediction as the data vector to test our models and pipelines.